// alife-logger/lib/biz.browser/fmp.js
var util = require("../util"), checkInterval = 500;

module.exports = function(e, n, t) {
    function r(e, n, t) {
        var i = 0, u = e.tagName;
        if ("SCRIPT" !== u && "STYLE" !== u && "META" !== u && "HEAD" !== u) {
            var c = e.children ? e.children.length : 0;
            if (c > 0) for (var a = e.children, l = c - 1; l >= 0; l--) i += r(a[l], n + 1, i > 0);
            if (i <= 0 && !t) {
                if (!(e.getBoundingClientRect && e.getBoundingClientRect().top < o)) return 0;
            }
            i += 1 + .5 * n;
        }
        return i;
    }
    function i(e) {
        for (var n = 1; n < e.length; n++) if (e[n].score < e[n - 1].score) return e.splice(n, 1), 
        i(e);
        return e;
    }
    var o = n.innerHeight || 0, u = [], c = null, a = 0;
    util.ext(e.prototype, {
        initFmpObserver: function(e) {
            var i = this;
            if (!i._conf || !i._conf.useFmp) return null;
            if (!n.MutationObserver) return util.warn("[retcode] first meaningful paint can not be retrieved"), 
            i.sendPerformance(), null;
            util.on(n, "beforeunload", function() {
                i.endObserving(0, !0);
            });
            var o = n.MutationObserver;
            return (c = new o(function() {
                !function(e) {
                    var n = Date.now() - e, i = t.querySelector("body");
                    if (i) {
                        var o = 0;
                        o += r(i, 1, !1), u.push({
                            score: o,
                            t: n
                        });
                    } else u.push({
                        score: 0,
                        t: n
                    });
                }(i._startTime);
            })).observe(document, {
                childList: !0,
                subtree: !0
            }), a = 1, i.onReady(function() {
                i.endObserving(e);
            }), c;
        },
        endObserving: function(e, n) {
            var t = this;
            if (c && a) if (n || !function(e, n) {
                var t = Date.now() - e;
                return !(t > n || t - (u && u.length && u[u.length - 1].t || 0) > 2 * checkInterval);
            }(t._startTime, e)) {
                c.disconnect(), a = 0, u = i(u);
                for (var r = null, o = 1; o < u.length; o++) if (u[o].t >= u[o - 1].t) {
                    var l = u[o].score - u[o - 1].score;
                    (!r || r.rate <= l) && (r = {
                        t: u[o].t,
                        rate: l
                    });
                }
                r && r.t > 0 && r.t < 36e5 ? t.sendPerformance({
                    fmp: r.t
                }) : t.sendPerformance();
            } else util.delay(function() {
                t.endObserving(e);
            }, checkInterval);
        }
    });
};
