// alife-logger/lib/util.js
Date.now = Date.now || function() {
    return new Date().getTime();
};

var SEQUENCE = Date.now(), noop = function() {}, getCwarn = function() {
    var t = "object" == typeof console ? console.warn : noop;
    try {
        var e = {
            warn: t
        };
        e.warn.call(e);
    } catch (n) {
        return noop;
    }
    return t;
}, util = {
    noop: noop,
    warn: getCwarn(),
    key: "__bl",
    win: "object" == typeof window && window.document ? window : undefined,
    regionMap: {
        cn: "https://arms-retcode.aliyuncs.com/r.png?",
        sg: "https://arms-retcode-sg.aliyuncs.com/r.png?",
        sg_2: "https://retcode-sg-lazada.arms.aliyuncs.com/r.png?",
        daily: "http://arms-retcode-daily.alibaba.net/r.png?",
        daily_2: "https://arms-retcode-daily.alibaba.net/r.png?",
        us: "https://retcode-us-west-1.arms.aliyuncs.com/r.png?"
    },
    defaultImgUrl: "https://arms-retcode.aliyuncs.com/r.png?",
    createObject: function(t) {
        if (Object.create) return Object.create(t);
        var e = function() {};
        return e.prototype = t, new e();
    },
    each: function(t, e) {
        var n = 0, r = t.length;
        if (this.T(t, "Array")) for (;n < r && !1 !== e.call(t[n], t[n], n); n++) ; else for (n in t) if (!1 === e.call(t[n], t[n], n)) break;
        return t;
    },
    safetyCall: function(t, e, n) {
        if ("function" != typeof t) return n;
        try {
            return t.apply(this, e);
        } catch (r) {
            return n;
        }
    },
    T: function(t, e) {
        var n = Object.prototype.toString.call(t).substring(8).replace("]", "");
        return e ? n === e : n;
    },
    filterByRule: function(t, e) {
        if (!t) return "";
        if (!e) return t;
        var n = this, r = n.T(e);
        return "Function" === r ? n.safetyCall(e, [ t ], t) : "Array" === r ? (this.each(e, function(e) {
            t = n.filterByRule(t, e);
        }), t) : "Object" === r ? t.replace(e.rule, e.target || "") : t.replace(e, "");
    },
    ignoreByRule: function(t, e) {
        if (!t || !e) return !1;
        if ((this.isString(e) || e.source || "Function" === this.T(e)) && (e = [ e ]), !this.isArray(e)) return util.warn("[arms] invalid rules of ignore config, (list of) String/RegExp/Funcitons are available"), 
        !1;
        for (var n, r = [], o = 0, i = e.length; o < i; o++) if (n = e[o], this.isString(n)) r.push(n.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, "\\$1")); else if (n && n.source) r.push(n.source); else if (n && "Function" === this.T(n) && !0 === this.safetyCall(n, [ t ], !1)) return !0;
        var a = new RegExp(r.join("|"), "i");
        return !!(r.length && a.test && a.test(t));
    },
    J: function(t) {
        if (!t || "string" != typeof t) return t;
        var e = null;
        try {
            e = JSON.parse(t);
        } catch (n) {}
        return e;
    },
    pick: function(t) {
        return 1 === t || 1 === Math.ceil(Math.random() * t);
    },
    verifyConfig: function(t) {
        if ("sample" in t) {
            var e = t.sample, n = e;
            e && /^\d+(\.\d+)?%$/.test(e) && (n = parseInt(100 / parseFloat(e))), 0 < n && 1 > n && (n = parseInt(1 / n)), 
            n >= 1 && n <= 100 ? t.sample = n : delete t.sample;
        }
        return t;
    },
    on: function(t, e, n, r, o) {
        return t.addEventListener ? (o = o || !1, t.addEventListener(e, function i(a) {
            r && t.removeEventListener(e, i, o), n.call(this, a);
        }, o)) : t.attachEvent && t.attachEvent("on" + e, function a(o) {
            r && t.detachEvent("on" + e, a), n.call(this, o);
        }), this;
    },
    off: function(t, e, n) {
        return n ? (t.removeEventListener ? t.removeEventListener(e, n) : t.detachEvent && t.detachEvent(e, n), 
        this) : this;
    },
    delay: function(t, e) {
        return -1 === e ? (t(), null) : setTimeout(t, e || 0);
    },
    ext: function(t) {
        for (var e = 1, n = arguments.length; e < n; e++) {
            var r = arguments[e];
            for (var o in r) Object.prototype.hasOwnProperty.call(r, o) && (t[o] = r[o]);
        }
        return t;
    },
    sub: function(t, e) {
        var n = {};
        return this.each(t, function(t, r) {
            -1 !== e.indexOf(r) && (n[r] = t);
        }), n;
    },
    uu: function() {
        for (var t, e, n = 20, r = new Array(n), o = Date.now().toString(36).split(""); n-- > 0; ) e = (t = 36 * Math.random() | 0).toString(36), 
        r[n] = t % 3 ? e : e.toUpperCase();
        for (var i = 0; i < 8; i++) r.splice(3 * i + 2, 0, o[i]);
        return r.join("");
    },
    seq: function() {
        return (SEQUENCE++).toString(36);
    },
    decode: function(t) {
        try {
            t = decodeURIComponent(t);
        } catch (e) {}
        return t;
    },
    encode: function(t, e) {
        try {
            t = e ? encodeURIComponent(t).replace(/\(/g, "%28").replace(/\)/g, "%29") : encodeURIComponent(t);
        } catch (n) {}
        return t;
    },
    serialize: function(t) {
        t = t || {};
        var e = [];
        for (var n in t) Object.prototype.hasOwnProperty.call(t, n) && t[n] !== undefined && e.push(n + "=" + this.encode(t[n], "msg" === n));
        return e.join("&");
    },
    checkAPI: function(t, e) {
        if (!t || "string" != typeof t) return !1;
        var n = /arms-retcode[\w-]*\.aliyuncs/.test(t);
        return !n && e && (n = /(\.png)|(\.gif)|(alicdn\.com)/.test(t)), !n;
    },
    checkAutoError: function(t) {
        return !(!t || !t.message) && !/failed[\w\s]+fetch/i.test(t.message);
    },
    cutUrlSearch: function(t) {
        return t && "string" == typeof t ? t.replace(/^(https?:)?\/\//, "").replace(/\?.*$/, "") : "";
    },
    removeUrlSearch: function(t) {
        return t && "string" == typeof t ? t.replace(/\?.*$/, "") : "";
    },
    createFakeToString: function(t) {
        return function() {
            return t + "() { [native code] }";
        };
    },
    checkSameOrigin: function(t, e) {
        if (!e || !t) return !1;
        var n = "//" + e.split("/")[2];
        return t === e || t.slice(0, e.length + 1) === e + "/" || t === n || t.slice(0, n.length + 1) === n + "/" || !/^(\/\/|http:|https:).*/.test(t);
    },
    getRandIP: function() {
        for (var t = [], e = 0; e < 4; e++) {
            var n = Math.floor(256 * Math.random());
            t[e] = (n > 15 ? "" : "0") + n.toString(16);
        }
        return t.join("");
    },
    getSortNum: function(t) {
        return t ? (t += 1) >= 1e3 && t <= 9999 ? t : t < 1e3 ? t + 1e3 : t % 1e4 + 1e3 : 1e3;
    },
    getRandNum: function(t) {
        return t && "string" == typeof t ? t.length < 5 ? this.getNum(5) : t.substring(t.length - 5) : this.getNum(5);
    },
    getNum: function(t) {
        for (var e = [], n = 0; n < t; n++) {
            var r = Math.floor(16 * Math.random());
            e[n] = r.toString(16);
        }
        return e.join("");
    },
    getCurDomain: function() {
        return location && location.hostname || "";
    },
    parseFetchHeaders: function(t) {
        if (!t) return {};
        var e = {};
        try {
            if ("function" == typeof t.keys) for (var n = t.keys(), r = n.next(); !r.done; ) {
                var o = r.value;
                e[o] = t.get(o), r = n.next();
            } else e = t;
        } catch (i) {
            e = {};
        }
        return e;
    },
    parseXhrHeaders: function(t) {
        if (!t && "string" != typeof t) return {};
        var e = {};
        try {
            var n = t.split("\r\n");
            e = n.reduce(function(t, e) {
                var n = e.split(": ");
                return t[n[0]] = n[1], t;
            }, {});
        } catch (r) {
            e = {};
        }
        return e;
    },
    getQuerys: function(t) {
        if (!t) return "";
        var e = {}, n = [], r = "", o = "";
        try {
            var i = [];
            if (t.indexOf("?") >= 0 && (i = t.substring(t.indexOf("?") + 1, t.length).split("&")), 
            i.length > 0) for (var a in i) r = (n = i[a].split("="))[0], o = n[1], e[r] = o;
        } catch (c) {
            e = {};
        }
        return e;
    },
    getFetchSnapshot: function(t, e, n) {
        var r, o;
        try {
            var i = (t && "string" != typeof t[0] ? t[0].url : t[0]) || "", a = (t && "string" != typeof t[0] ? t[0] : t[1]) || {}, c = "POST" === a.method.toUpperCase() ? a.body : this.getQuerys(i);
            r = {
                originApi: i,
                method: a.method || "unknown",
                params: c,
                response: e || "",
                reqHeaders: this.parseFetchHeaders(a.headers || null),
                resHeaders: this.parseFetchHeaders(n)
            }, o = "function" == typeof encodeURIComponent && JSON && encodeURIComponent(JSON.stringify(r)) || "{}";
        } catch (s) {
            o = "{}";
        }
        return o;
    },
    getXhrSnapshot: function(t, e, n) {
        if (!t || !e || !n) return {};
        var r, o;
        try {
            r = {
                originApi: t,
                method: e,
                params: this.getQuerys(t),
                response: n.responseText || n.responseXml || "",
                reqHeaders: {},
                resHeaders: this.parseXhrHeaders("function" == typeof n.getAllResponseHeaders && n.getAllResponseHeaders() || "")
            }, o = "function" == typeof encodeURIComponent && JSON && encodeURIComponent(JSON.stringify(r)) || "{}";
        } catch (i) {
            o = "{}";
        }
        return o;
    },
    isFunction: function(t) {
        return "function" == typeof t;
    },
    isPlainObject: function(t) {
        return "[object Object]" === Object.prototype.toString.call(t);
    },
    isString: function(t) {
        return "[object String]" === Object.prototype.toString.call(t);
    },
    isArray: function(t) {
        return "[object Array]" === Object.prototype.toString.call(t);
    },
    joinRegExp: function(t) {
        for (var e, n = [], r = 0, o = t.length; r < o; r++) e = t[r], this.isString(e) ? n.push(e.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, "\\$1")) : e && e.source && n.push(e.source);
        return new RegExp(n.join("|"), "i");
    },
    reWriteMethod: function(t, e, n) {
        if (null !== t) {
            var r = t[e];
            t[e] = n(r);
        }
    }
};

module.exports = util;
