// alife-logger/lib/biz.eapp/hook.js
module.exports = function(e) {
    var t = require("../util"), i = null;
    return t.ext(e.prototype, {
        addHook: function() {
            return this.isHookInstantiated ? this : (function() {
                var e = this;
                if ("undefined" != typeof dd && dd && "function" == typeof dd.httpRequest) {
                    var s = dd.httpRequest;
                    i = s;
                    var a = Object.getOwnPropertyDescriptor(dd, "httpRequest");
                    a && a.writable && (dd.httpRequest = function(i) {
                        var a = new Date().getTime(), n = i;
                        if (i && i.url) {
                            var o = t.cutUrlSearch(i.url);
                            if (!t.checkAPI(o, !0)) return s.call(dd, n);
                            var r, d, c = n && n.headers;
                            c && "object" == typeof c || (c = {});
                            var u = {};
                            if (e.getConfig("enableLinkTrace")) {
                                var f = c["EagleEye-pAppName"];
                                if (r = c["EagleEye-TraceID"], d = c["EagleEye-SessionID"], r || (r = e.getTraceId()["EagleEye-TraceID"], 
                                u["EagleEye-TraceID"] = r), d || (d = e.getSessionId()["EagleEye-SessionID"], u["EagleEye-SessionID"] = d), 
                                !f) {
                                    var l = e.getConfig("pid");
                                    u["EagleEye-pAppName"] = l;
                                }
                            }
                            var p = {
                                success: function(t) {
                                    "function" == typeof i.success && i.success(t);
                                    var s = new Date().getTime();
                                    e.api({
                                        api: o,
                                        success: !0,
                                        time: s - a,
                                        code: t && t.status || 200,
                                        begin: a,
                                        traceId: r,
                                        sid: d
                                    });
                                },
                                fail: function(t) {
                                    "function" == typeof i.fail && i.fail(t);
                                    var s = new Date().getTime(), n = "";
                                    t && t.status && (n = t.status);
                                    var c = "";
                                    t && t.body && (c = (c = JSON.stringify(t.body)).substring(0, 1e3)), e.api({
                                        api: o,
                                        success: !1,
                                        time: s - a,
                                        code: n,
                                        msg: c,
                                        begin: a,
                                        traceId: r,
                                        sid: d
                                    });
                                },
                                headers: t.ext({}, c, u)
                            };
                            n = t.ext({}, n, p);
                        }
                        return s.call(dd, n);
                    });
                }
            }.call(this), this.isHookInstantiated = !0, this);
        },
        removeHook: function() {
            return this.isHookInstantiated ? (function() {
                if ("undefined" != typeof dd && dd && i) {
                    var e = Object.getOwnPropertyDescriptor(dd, "httpRequest");
                    e && e.writable && (dd.httpRequest = i), i = null;
                }
            }.call(this), this.isHookInstantiated = !1, this) : this;
        },
        initHook: function() {
            return this.hasInitHook ? this : (this.getConfig("disableHook") || this.addHook(), 
            this.hasInitHook = !0, this);
        }
    }), e;
};
