// alife-logger/.eslintrc.js
// # ----------------------------------
// # eslint rules
// # @author: Pluto <huazhi.chz@alibaba-inc.com>
// # @create: 2017/06/06
// # ----------------------------------

module.exports = {
    root: true,
    env: {
        browser: true,
        node: false,
        es6: false,
        commonjs: true,
        jquery: false
    },
    plugins: [],
    extends: ['eslint:recommended'],
    parserOptions: {
        sourceType: 'module',
        ecmaVersion: 5,
        ecmaFeatures: {
            jsx: false,
            experimentalObjectRestSpread: false
        }
    },
    globals: {
        __bl: true,
        define: true,
        require: true,
        module: true,
        chrome: true,
        WXEnvironment: true,
        dd: true,
        my: true,
        wx: true,
        getCurrentPages: true,
    },
    rules: {
        'no-unused-vars': 'warn',
        'no-console': ['warn', { allow: ['warn', 'error'] }],
        indent: ['warn', 4, {SwitchCase: 1}],
        quotes: ['error', 'single'],
        semi: ['warn', 'always', {omitLastInOneLineBlock: true}],
        'default-case': 'error',
        eqeqeq: ['error', 'always'],
        'no-warning-comments': ['error', { 'terms': ['todo', 'fixme', 'xxx'], 'location': 'start' }],
        'comma-spacing': ['error', { 'before': false, 'after': true }],
        'key-spacing': ['warn', {beforeColon: false, afterColon: true}],
        'space-before-blocks': ['warn', 'always']
    }
};
