// alife-logger/lib/biz.wx/clazz.js
var util = require("../util"), MiniProgramLogger = require("../miniProgramLogger"), ARMS_STORAGE_MINIPROGRAM_WX_UID_KEY = "ARMS_STORAGE_MINIPROGRAM_WX_UID_KEY", WXLogger = function(e) {
    return MiniProgramLogger.call(this, e), this;
};

WXLogger.prototype = util.createObject(MiniProgramLogger.prototype), util.ext(MiniProgramLogger._root.dftCon, {
    sendRequest: function(e, t) {
        if ("undefined" != typeof wx && wx && "function" == typeof wx.request) try {
            var r, o = "GET";
            t && (o = "POST", r = JSON.stringify(t)), wx.request({
                url: e,
                method: o,
                data: r,
                fail: function(e) {
                    util.warn("[arms] sendRequest fail", e);
                }
            });
        } catch (n) {
            util.warn("[arms] error in conf sendRequest", n);
        }
    },
    getCurrentPage: function() {
        if ("function" == typeof getCurrentPages) try {
            var e = getCurrentPages() || [], t = e[e.length - 1];
            return t && t.route || null;
        } catch (r) {
            util.warn("[arms] error in conf getCurrentPage", r);
        }
    }
}), util.ext(WXLogger.prototype, {
    constructor: WXLogger,
    _super: MiniProgramLogger,
    autoSetCommonInfo: function() {
        this.setCommonInfo({
            app: "mini_wx"
        }), this.autoSetSystemInfo(), this.autoSetNetworkType(), this.autoSetUid();
    },
    autoSetUid: function() {
        if (this._conf && this._conf.uid) this.setCommonInfo({
            uid: this._conf.uid
        }); else if ("undefined" != typeof wx && wx && "function" == typeof wx.getStorageSync) try {
            var e = wx.getStorageSync(ARMS_STORAGE_MINIPROGRAM_WX_UID_KEY);
            if (e && "string" == typeof e) this.setCommonInfo({
                uid: e
            }); else if ("function" == typeof wx.setStorageSync) {
                var t = util.uu();
                wx.setStorageSync(ARMS_STORAGE_MINIPROGRAM_WX_UID_KEY, t), this.setCommonInfo({
                    uid: t
                });
            }
        } catch (r) {
            util.warn("[arms] error in autoSetUid", r);
        }
    },
    autoSetSystemInfo: function() {
        if ("undefined" != typeof wx && wx && "function" == typeof wx.getSystemInfoSync) try {
            var e = wx.getSystemInfoSync();
            "object" == typeof e && this.setCommonInfo({
                sr: (e.screenWidth || 0) + "x" + (e.screenHeight || 0),
                vp: (e.windowWidth || 0) + "x" + (e.windowHeight || 0),
                dpr: e.pixelRatio,
                ul: e.language
            });
        } catch (t) {
            util.warn("[arms] error in autoSetSystemInfo", t);
        }
    },
    autoSetNetworkType: function() {
        var e = this;
        if ("undefined" != typeof wx && wx && "function" == typeof wx.getNetworkType) try {
            wx.getNetworkType({
                success: function(t) {
                    t && "string" == typeof t.networkType && e.setCommonInfo({
                        ct: t.networkType
                    });
                },
                fail: function(e) {
                    util.warn("[arms] autoSetNetworkType getNetworkType fail", e);
                }
            });
        } catch (t) {
            util.warn("[arms] error in autoSetNetworkType", t);
        }
    },
    hookApp: function(e) {
        var t = this, r = {
            onError: function(r) {
                var o = 1 === arguments.length ? [ arguments[0] ] : Array.apply(null, arguments), n = e.onError;
                try {
                    if (r && "object" == typeof r && t.error(r), r && "string" == typeof r) {
                        var i = r.split("\n"), g = "", a = "";
                        i.length > 1 && (g = i[0] && i[0].length < 100 ? i[0] : i[0].substring(0, 100), 
                        a = i[1]), t.error({
                            name: g,
                            message: a || r,
                            stack: r
                        });
                    }
                } catch (r) {
                    util.warn("[arms] error in hookApp:onError", r);
                }
                if ("function" == typeof n) return n.apply(this, o);
            }
        };
        return util.ext({}, e, r);
    }
}), require("./hook")(WXLogger);

var behavior = require("./behavior");

"undefined" != typeof wx && wx && behavior(WXLogger, wx);

var singleton = null, generator = function(e) {
    return singleton || (singleton = new WXLogger(e || {})), singleton;
};

WXLogger.createExtraInstance = function(e) {
    e && "object" == typeof e ? (e.disableHook = !0, e.behavior = !1) : e = {
        disableHook: !0,
        behavior: !1
    };
    return new WXLogger(e);
}, WXLogger.init = generator, WXLogger.singleton = generator, WXLogger._super = MiniProgramLogger, 
WXLogger._root = MiniProgramLogger._root, MiniProgramLogger.WXLogger = WXLogger, 
module.exports = WXLogger;
