// alife-logger/lib/biz.weex/env.js
var util = require("../util"), wx = "object" == typeof WXEnvironment && WXEnvironment.weexVersion ? WXEnvironment : undefined, win = "object" == typeof window && window.addEventListener ? window : undefined, isH5 = !wx || "Web" === wx.platform;

module.exports = {
    wx: wx,
    win: win,
    isH5: isH5,
    getSr: function() {
        return isH5 ? screen.width + "x" + screen.height : wx.deviceWidth + "x" + wx.deviceHeight;
    },
    getH5Page: function(e, n) {
        var t = location, i = t.host + t.pathname;
        return util.filterByRule(n ? i.toLowerCase() : i, e);
    },
    getPvInfo: function() {
        var e = isH5 ? document : undefined, n = isH5 ? win.devicePixelRatio || 1 : 1;
        return isH5 ? {
            begin: Date.now(),
            dt: e.title,
            dl: location.href,
            dr: e.referrer,
            dpr: n.toFixed(2),
            de: (e.characterSet || e.defaultCharset || "").toLowerCase(),
            ul: e.documentElement.lang
        } : {
            begin: Date.now()
        };
    }
};
